python/services/dataservice/dmod/dataservice/service.py

```diff
@@ -1025,4 +830,206 @@ async def perform_checks_for_job(self, job: Job) -> bool:
+                        job_ds_user.link_to_dataset(ds)
+
+                    # Initialize dataset Docker volumes required for a job
+                    logging.debug('Initializing any required S3FS dataset volumes for {}'.format(job.job_id))
+                    self._docker_s3fs_helper.init_volumes(job=job)
+                except Exception:
+                    job.set_status_step(JobExecStep.DATA_FAILURE)
+                    self._job_util.save_job(job)
+                    continue
+                finally:
+                    self._count.release()
+
+                job.set_status_step(JobExecStep.AWAITING_SCHEDULING)
+                self._job_util.save_job(job)
+
+            # Also, unlink usage for any previously existing, job-based users for which the job is no longer active ...
+            self._unlink_finished_jobs(ds_users=prior_users)
+
+            self._job_util.unlock_active_jobs(lock_id)
+            await asyncio.sleep(5)
+
+    def _unlink_finished_jobs(self, ds_users: Dict[UUID, DatasetUser]) -> Set[UUID]:
+        """
+        Unlink dataset use for any provided job-based users for which the job is no longer active.
+
+        Unlink dataset use for any of the given dataset users that are specifically ::class`JobDatasetUser` and are
+        associated with a job that is not in the set of active jobs.
+
+        Parameters
+        ----------
+        ds_users: Dict[UUID, DatasetUser]
+            A mapping of applicable dataset user objects, mapped by uuid; for the subset of interest, which are
+            ::class`JobDatasetUser` instances, the key will also be the id of the associated job (as a ::class:`UUID`).
+
+        Returns
+        -------
+        Set[UUID]
+            Set of the UUIDs of all users associated with finished jobs for which unlinking was performed.
+        """
+        finished_job_users: Set[UUID] = set()
+        active_job_ids = {UUID(j.job_id) for j in self._job_util.get_all_active_jobs()}
+        for user in (u for uid, u in ds_users.items() if isinstance(u, JobDatasetUser) and uid not in active_job_ids):
+            for ds in (self._managers.known_datasets()[ds_name] for ds_name in user.datasets_and_managers):
+                user.unlink_to_dataset(ds)
+            finished_job_users.add(user.uuid)
+        return finished_job_users
+
+def _get_ds_users(managers: DatasetManagerCollection) -> Dict[UUID, DatasetUser]:
+    """
+    Get dataset users from all associated managers of the service.
+
+    Returns
+    -------
+    Dict[UUID, DatasetUser]
+        Map of all ::class:`DatasetUser` from all associated managers of the service, keyed by the UUID of each.
+    """
+    return {uuid: mgr.get_dataset_user(uuid)
+            for _, mgr in managers.managers() for uuid in mgr.get_dataset_user_ids()}
+
+class Count:
+    """
+    Unbound Semaphore-like class for counting acquisitions and releases. Unlike a semaphore, `acquire()` nor `release()`
```

Comment:
Can you include an `Examples:` tag to show its expected use as a member in another object? Just something to show what to expect.